### PR TITLE
ENH: Adding callbacks to Norms for update signals

### LINKED ISSUE #4387: make `Normalize` objects notifiy scalar-mappables on changes
Currently just changing the limit of a normalizer will not invalidate the caches in `AxesImage` so the figure will not update to reflect the changed limits.  The reason you would want to do this is that by sharing a `Normalize` instance between multiple scalar mappable objects you can stay synced similar to `sharex` and `sharey`.

### LINKED ISSUE #17052: Colorbar update error with clim change in multi_image.py example
### Bug report

**Bug summary**

In the multi_image example, the colorbar is responding correctly to `set_clim` only when called on the image to which the colorbar is directly attached.   

**Code for reproduction**
This is just the example, https://matplotlib.org/gallery/images_contours_and_fields/multi_image.html, with manipulation of the clim at the bottom.
```python
from matplotlib import colors
import matplotlib.pyplot as plt
import numpy as np

np.random.seed(19680801)
Nr = 3
Nc = 2
cmap = "cool"

fig, axs = plt.subplots(Nr, Nc)
fig.suptitle('Multiple images')

images = []
for i in range(Nr):
    for j in range(Nc):
        # Generate data with a range that varies from one plot to the next.
        data = ((1 + i + j) / 10) * np.random.rand(10, 20) * 1e-6
        images.append(axs[i, j].imshow(data, cmap=cmap))
        axs[i, j].label_outer()

# Find the min and max of all colors for use in setting the color scale.
vmin = min(image.get_array().min() for image in images)
vmax = max(image.get_array().max() for image in images)
norm = colors.Normalize(vmin=vmin, vmax=vmax)
for im in images:
    im.set_norm(norm)

fig.colorbar(images[0], ax=axs, orientation='horizontal', fraction=.1)


# Make images respond to changes in the norm of other images (e.g. via the
# "edit axis, curves and images parameters" GUI on Qt), but be careful not to
# recurse infinitely!
def update(changed_image):
    for im in images:
        if (changed_image.get_cmap() != im.get_cmap()
                or changed_image.get_clim() != im.get_clim()):
            im.set_cmap(changed_image.get_cmap())
            im.set_clim(changed_image.get_clim())


for im in images:
    im.callbacksSM.connect('changed', update)

images[1].set_clim(1e-9, 2e-8)
fig.savefig('ax1_bad.png')
images[0].set_clim(1e-9, 2e-8)
fig.savefig('ax0_good.png')

```

**Actual outcome**
ax1_bad.png:

ax1_bad

**Expected outcome**
ax0_good.png:

ax0_good

**Matplotlib version**

  * Operating system: OSX
  * Matplotlib version:  3.1.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX, agg
  * Python version: 3.7

### PR DESCRIPTION
**Pull request by greglucas**
## PR Summary

This adds a callback registry to Norm instances that can be connected to by other objects to be notified when the Norm is updated. This is particularly relevant for ScalarMappables to be notified when the vmin/vmax are changed on the Norm. With this update, the axis is now immediately registered as stale in the following example.

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
norm = plt.Normalize(-2, 2)
ax.imshow([[0, 1], [2, -1]], norm=norm)

norm.vmax = 0.1
print(ax.stale)
```

Closes #4387
Fixes #17052

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [N/A] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by greglucas** [comment]
I _think_ I have this down to one failing test now, but I'm not entirely sure why its failing. It is in AxesGrid and looks like the cbar axis isn't getting cleared or set up properly here: https://github.com/matplotlib/matplotlib/blob/a510e32a37a8b4e0bc4e8384df20c7b45abd9913/lib/mpl_toolkits/tests/test_axes_grid.py#L42-L45

The only major change I made to colorbar was to deepcopy the norm, but that doesn't seem like the error here. Here is the image produced with this new PR.

imagegrid_cbar_mode

**Review by jklymak — DISMISSED** [review]
Sorry, I'll block on this feature, and suggest #4387 be closed.  

In general I am against magic callbacks like this.  I think an artist should reflect the state of the artist at creation time, or when it is explicitly acted upon, not at draw time, and possibly based on actions applied to other artists.  If you need to keep two artists limits in sync, you should explicitly do so, and not rely on a black box callback.  

I appreciate there was a push to make the library cache more things until draw time, and maybe if the whole library had been designed with that goal from the start that would make sense.  But as-is, implicit state is simply confusing and leads to unexpected behaviour. 

Of course happy to unblock if there is a real strong case for why this is needed, and can't be done at the user level.

**Comment by greglucas** [comment]
I think I agree with you...? This isn't caching until draw-time, it is just creating an update mechanism that will automatically trigger the re-drawing chain (norm -> scalar-mappable -> artist). Someone still has to explicitly modify the norm.

This was motivated by me wanting to have interactivity with the colorbars and panning/zooming on that axis, see here: https://github.com/matplotlib/matplotlib/pull/19515

Basically, I was trying to bring things a little more in-sync throughout the pipeline of the scalar-mappable updates to not force a user to call `fig.draw()` to update the image. See here for other scalar-mappable cleanup work as well:
https://github.com/matplotlib/matplotlib/pull/19552

**Comment by jklymak** [comment]
> This was motivated by me wanting to have interactivity with the colorbars and panning/zooming on that axis

I see.  But that feature should also be discussed I guess?  Are you suggesting if I zoom on a colorbar, that should change the mappable somehow?   How?  And why is that a useful thing to do?  I guess I could imagine the converse, where you set the limits of the colorbar to change vmin and vmax, but that would require a text box or something, rather than zooming.  

But here, you are pushing it further, and not only changing the mappable that belongs to the colorbar, but changing the norm and wanting the changes to propagate to other mappables that may have the "same" norm.  

I guess I'm arguing we should be doing exactly the opposite, and making the norms a copy when added to the mappable, and hence a change to one mappable does *not* affect the others.  But maybe I haven't though that through and many folks are making a single norm, and sharing it, and then expecting to manipulate the norm and have it show up on the next draw.  I guess the PR would benefit from a bit more context of what the current state is, and why this is an improvement?

**Comment by dopplershift** [comment]
@jklymak The current state is that this:
```python
fig, ax = plt.subplots()
norm = plt.Normalize(-2, 2)
cmap = copy.copy(plt.get_cmap('BuGn'))
ax.imshow(data, norm=norm, cmap=cmap)
norm.vmax = 0.1
norm.vmin = -0.1
cmap.set_over('red')
```
plots with a norm from 0.1 to -0.1, and a colormap with an "over" color of red.

I'm not a fan of having the API make a copy of an object. I don't know of anywhere else in the API that we do that (e.g. `add_artist`).

**Comment by jklymak** [comment]
OK, sure, but without running it, are you sure what this does?  

```python
import matplotlib.pyplot as plt
import numpy as np
import copy 

fig, axs = plt.subplots(1,2)
norm = plt.Normalize(-2, 2)
data = np.linspace(-1, 1, 900).reshape(30, 30)
cmap = copy.copy(plt.get_cmap('BuGn'))
axs[0].imshow(data, norm=norm, cmap=cmap)
norm.vmax = 0.1
norm.vmin = -0.1
cmap.set_over('red')

im = axs[1].imshow(data, norm=norm, cmap=cmap)
fig.colorbar(im, ax=axs)

im.set_clim(-0.3, 0.3)

plt.show()
```

**Comment by greglucas** [comment]
I will say what I _think_ should happen without running it. Everything is using the same norm, so everything should be in-sync with the final updates that were made before the `plt.show()` call. So, I'd expect the norm to be (-0.3, 0.3), and the images to show red at any values over 0.3.

If you run this in an interactive session, currently when you modify the norm `norm.vmax = 0.1` nothing will happen. However, with the new callback added in this PR, the norms would signal a `stale` figure and trigger a redraw automatically. This could be attached to a text/slider widget as you mentioned above which would be a great way to change the color limits too.

I'll try to add some more context to the original PR summary

**Comment by jklymak** [comment]
OK, thanks @greglucas and @dopplershift  - I thought about this some and agree with both of you.  This sort of behaviour *can* be a bit mysterious - I wouldn't expect the user to know that the clim on an object is linked up to a global norm instance, but I agree with you that this is how it is, and that such behaviour can have uses....

**Comment by tacaswell** [comment]
As the author of the linked issue I am (still) in favor of this :)

The problem this is trying to solve is that we do have a number of caches (for performance reasons) that need to be invalidated when we change the norm.  If you do it one way (through `set_clim`) (which invalidates the cache) and the other is to set the 

```
fig, axs = plt.subplots(1,2)
norm = plt.Normalize(-2, 2)
data = np.linspace(-1, 1, 900).reshape(30, 30)
cmap = copy.copy(plt.get_cmap('BuGn'))
axs[0].imshow(data, norm=norm, cmap=cmap)
# norm.vmax = 0.1
# norm.vmin = -0.1

norm = plt.Normalize(-2, 2)
axs[1].imshow(data, norm=norm, cmap=cmap)
print(f'{norm.vmin}, {norm.vmax}')
plt.pause(5)
norm.vmax = 0.1
norm.vmin = -0.1

print(f'{norm.vmin}, {norm.vmax}, {axs[1].images[0].get_clim()}')
# at this point the two Axes will look identical despite reporting different clims!
plt.show()
# if you force a full redraw by e.g. resizing the window they will snap to be different.
```

This is adding an explicit coupling (via the callback) to get rid of an implicit coupling (with hysteresis and dependencies on exactly where draws are!).

I have not run the code on this PR, but I think it will fix this.

We will need to check than `obj.changed()` is relatively cheap or de-bounce it somehow in `im.set_clim()` (which will not trigger this 3 times: 2 each from vmin/vmax and once from an explicit call to `self.changed()` in `set_clim`.

**Comment by greglucas** [comment]
@jklymak, just to be clear there is **no** global Norm here. This is a specific norm object and you _can_ share that object between images to keep them in sync. But, if you put `plt.Normalize()` on one image and `plt.Normalize()` on a second image, those will be two separate norms that can be modified independently and this won't sync those.

@tacaswell, I got rid of the 3rd `on_changed` call in set_clim. There will still be two calls (vmin and vmax) as you mention, but as far as I can tell these seem like cheap updates that update `self.stale` and notify other listeners which also appear to be cheap at first glance.

**Comment by jklymak** [comment]
> @jklymak, just to be clear there is **no** global Norm here. 

No I understand, but there is a top-level `norm` object that the artist methods can manipulate (i.e. via `artist.set_clim`) and it is not necessarily clear that would affect a norm buried on the artist and thus potentially affect other artists.  i.e. you can easily imagine a user wanting two images to have log-norm scaling, but not necessarily the same vmin and vmax and doing:

```python
norm = plt.LogNorm()
im1 = ax1.imshow(X1, norm=norm)
im2 = ax2.inshow(X2, norm=norm)
im2.set_clim(1e-3, 1e5)
```
and be surprised that this affects `im1`.  i.e they weren't thinking of the norm holistically as scale, vmin, vmax,, but just as a way to get logarithmic scaling.

**Comment by greglucas** [comment]
That makes sense, but that is the case regardless of this PR (both images are modified on current master), and I think not too bad to explain to someone not to re-use an object like that but rather just create separate `LogNorm`s.

File: lib/matplotlib/contour.py
Comment (jklymak) [inline lib/matplotlib/contour.py:1030]: this behaviour seems new - can you clarify why you need this?  The explanation below didn't help me understand why this is needed now.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (tacaswell) [inline lib/mpl_toolkits/tests/test_axes_grid.py:49]: Did we used to rely on a race condition in drawing to make this test work?

I'm also not seeing in the above code how they ended up coupled.

File: lib/matplotlib/colors.py
Comment (greglucas) [reply lib/matplotlib/colors.py:1345]: Thanks, this actually caught that I forgot to add the update check to `CenteredNorm()`

File: lib/matplotlib/contour.py
Comment (greglucas) [reply lib/matplotlib/contour.py:1030]: This is actually to keep the behavior the same as before. I kept the commits that fix the tests in with messages to try and explain why I did this. (See the second and third commit messages)

At a high level though, this `changed()` overrides the one in ScalarMappable and we need to make sure that we are "changing" in the appropriate place. This was a little odd because contour does some extra things with cvalues and levels that aren't needed on any of the other mappables, so we need to autoscale and adjust the limits in the right order.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py:49]: This one took me a long time to figure out that this one-line fix is all that is needed. If I remove the deepcopy, the scales on the colorbar axes seem to stick around or not get updated appropriately? There are some extra ticks floating around and some spines don't get fully removed. I think there is some state that isn't fully getting cleared in `axes_grid` with the way the colorbars are handled there.

I don't think this is the most ideal test with this comment about clearing and overwriting portions of the axes.
```python
        # In each row/column, the "first" colorbars must be overwritten by the
        # "second" ones.  To achieve this, clear out the axes first.
```

Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: This test bit me again... I just added some new code to connect/disconnect the callbacks when changing the norm on a ScalarMappable that already exists. Now, I'm able to remove the `ax.cax.cla()` call, but I need a brand new ScalarMappable for only this test. I think there is something with the colorbars trying to modify old state here, which is why creating a new instance that has no sharing is what we need.

I still don't 100% follow this test code and haven't taken the time to really investigate it in depth and determine whether this test could be rewritten, or if something else in colorbars should be updated.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (jklymak) [inline lib/mpl_toolkits/tests/test_axes_grid.py:49]: I think @anntzer worked on this last, trying to make these colorbar-like objects the same as main-library colorbars.

File: lib/matplotlib/tests/test_colors.py
Comment (jklymak) [inline lib/matplotlib/tests/test_colors.py:1435]: Are these asserts enough? I guess it would be nice to know that the `stale` does something to the `sm`.  Maybe a figure equivalency test that tests setting the norm directly versus changing the norm after creation and proving that the two figures are the same?  It would be nice to see that the whole chain works.  See `check_figures_equal` decorators for a way to do this w/o adding extra figures to the repo.

Comment (jklymak) [reply lib/matplotlib/tests/test_colors.py:1435]: BTW, sorry this languished.  Feel free to ping me and I'l try and shepherd through...

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py:49]: This still feels like there is something different about `axes_grid` that I am missing and we shouldn't need to deepcopy the norm on the mappable... My guess is there is possibly a locator or something similar attached to `ax.cax` that needs to be cleared/reset on the `ax.cax.cla()` call above, but I haven't been able to find it while debugging and have spent quite a lot of time on this.

Is there a way to step through a debugger session while keeping the plot window open to see how things move around while stepping into functions?

Comment (jklymak) [reply lib/mpl_toolkits/tests/test_axes_grid.py:49]: Not that I know of, but I never use a debugger (probably a bad idea).

File: lib/matplotlib/tests/test_colors.py
Comment (jklymak) [reply lib/matplotlib/tests/test_colors.py:1435]: Somewhat the opposite - if I give something a hard time, but realize I was wrong, I feel some compunction to help correct my mistake.

For the test, I think using the reference figure with the norm set initially, and the compare figure with the norm modified after the fact to match the reference figure should work?  i.e basically what you have here, but a reference figure that starts with the `vcenter=1`.

Comment (greglucas) [reply lib/matplotlib/tests/test_colors.py:1441]: :+1: I added a quick figure test comparison now.

**Review by jklymak — APPROVED** [review]

**Comment by jklymak** [comment]
@tacaswell I'll try and pop this back on your PR review queue...

File: lib/matplotlib/contour.py
Comment (tacaswell) [reply lib/matplotlib/contour.py:1064]: The joys of making things a bit lazier!

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (tacaswell) [reply lib/mpl_toolkits/tests/test_axes_grid.py:49]: I'm still deeply confused by this test.  It seems that it intentionally over-rides the axes so I guess the point is to show that 'cax' for both of the axes in a given row / column?


imagegrid_cbar_mode

I am not sure what the expected behavior here actual is to sort out how big of a change this is...

Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: I am very confused as well. I _think_ what is happening is that on the first pass of colorbar addition to the cax, some locators/formatters get updated and set due to the new callback. Then on the second addition of the colorbar, that state is still hanging around. Rather than deepcopying, I switched the test to remove the callback from the colorbar.

The difference with/without the final commit is that the second colorbar axis has ticklabels showing up. I'm not exactly sure where this state creeps in, so this is a hard one to trace down.
imagegrid_cbar_mode-failed-diff

Comment (jklymak) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: @anntzer worked on this last in imageGrid, so it would be very helpful if he could describe what is going on here.

Comment (anntzer) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: From a quick look, this seems to arise (at least to some degree) because remove_text doesn't know how to hand axisartist-axes, which require a different API for removing ticklabels (because because).
(It's not clear to me why it worked before, though.)

Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: Thanks for looking into this @anntzer! I'll try and dig a little more into the axisartist/remove_text and see if there is a quick fix in there. Otherwise, I think a better workaround would likely be to move my fix in the test internal to the axes_grid colorbar routine and put a TODO in there for adding that capability back at a later point.

**Comment by greglucas** [comment]
I think this is ready for a final review now. I took the easy way out and removed the callback from all axes_grid colorbars for now to make the old test pass. I squashed down the commits now and left the relevant notes in the commit message.

File: lib/mpl_toolkits/axes_grid1/axes_grid.py
Comment (jklymak) [inline lib/mpl_toolkits/axes_grid1/axes_grid.py:37]: This seems like it needs investigation.  You want to add a callback machinery, which I am somewhat leery of because I don't like two-way state. Here it breaks down, so you just remove it, which seems like skirting the problem.    I'm not saying axes_grid1 is doing the right thing, but it seems like a red flag to me that should be understood.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (jklymak) [inline lib/mpl_toolkits/tests/test_axes_grid.py:47]: Why did you drop the reference?

File: lib/matplotlib/tests/test_image.py
Comment (greglucas) [reply lib/matplotlib/tests/test_image.py:1020]: The previous warning was:
```bash
matplotlib/lib/matplotlib/image.py:442: UserWarning: Warning: converting a masked element to nan.
  vmid = np.float64(self.norm.vmin) + dv / 2
```
If I print out the previous `im.norm.vmin, im.norm.vmax`, they were both `np.ma.masked`, which stems from passing in a fully masked array (`np.max(np.ma.array([1, 2], mask=[True, True]))` returns `masked`), so here we are calling `np.float64(np.ma.masked)`. I don't think we want `masked` as vmin/vmax, so this perhaps even helps here... The vmin/vmax turn into `(0, 0)` now.

Note that the image doesn't change at all, just no warning about norm vmin/vmax anymore.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py:47]: It is never used, so when I was testing I'm guessing I got rid of it and didn't realize it was dropped at some point.

File: doc/users/next_whats_new/callbacks_on_norms.rst
Comment (tacaswell) [inline doc/users/next_whats_new/callbacks_on_norms.rst:7]: This needs a note about what the key in the registry will be.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (anntzer) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: OK, I figured out this problem.  This is basically due to the interaction between two points.

First, as discussed during the call, cla() doesn't call Colorbar.remove(), which means that the callback of the first image doesn't get deleted, which it should.  Making cla() actually go through each and every artist to call cla() on it may be too expensive, but at least perhaps we can check if there's any colorbar on the axes and properly remove() them.

Still, even knowing that, it was unclear why the first image would overwrite the second colorbar.  The reason is because the sequence of events was the following:
- instantiate first image, create associated colorbar.
- instantiate second image, create associated colorbar.
- actually walk the draw tree and draw the images... which has the effect of calling _ImageBase._make_image, which contains the following snippet:
```python
                # we have re-set the vmin/vmax to account for small errors
                # that may have moved input values in/out of range
                s_vmin, s_vmax = vrange
                if isinstance(self.norm, mcolors.LogNorm) and s_vmin <= 0:
                    # Don't give 0 or negative values to LogNorm
                    s_vmin = np.finfo(scaled_dtype).eps
                with cbook._setattr_cm(self.norm,
                                       vmin=s_vmin,
                                       vmax=s_vmax,
                                       ):
                    output = self.norm(resampled_masked)
```
so here s_vmin and s_vmax were recomputed to take interpolation into account.  Unfortunately, this means they may be slightly different due to the rescaling and unscaling, which can be checked by printing `s_vmin - A.min(), s_vmax - A.max()`.  One then sees that there's indeed a tiny (1e-16) difference for the first image's norm, which means it'll say, oh, I need to redraw the colorbar; but there's no difference(!) for the second image's norm, so it doesn't redraw itself.  A quickfix to confirm that this works is that forcing `s_vmin = A.min(); s_vmax = A.max()` here fixes the issue (but would cause other problems with over/under interaction with resampling; I suspect that the "correct" fix is to temporarily disable emission of callbacks by `self.norm` just for the call to `_setattr_cm`.

That was nasty :p

**Comment by timhoffm** [comment]
If we go with a callback registry, it should be named `callbacks`. That's the name used for the registry in `Axes`, `Axis` and `Figure`. Only `ScalarMappable` uses `callbacksSM` for no obvious reason.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: That looks like it did the trick. I disconnected from the callback before this and reconnected after it. I wonder if more fundamentally there should be something within the context manager to remove the callbacks on enter/exit...

**Comment by greglucas** [comment]
Also fixes #17052 which I just found.

File: lib/mpl_toolkits/tests/test_axes_grid.py
Comment (anntzer) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: I guess we could add a `disabling_callbacks()` context manager on CallbackRegistry?
```python
with self.norm.callbacks.disabling_callbacks(), cbook._setattr_cm(self.norm, ...): ...
```

Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: Yeah, that looks pretty clean to me. I'm not sure how much more use it would have in more than this one situation though, so is it worth adding or not...

Lets keep it as a separate issue and implement it later.

Comment (greglucas) [reply lib/mpl_toolkits/tests/test_axes_grid.py]: Rebased and updated this with the blocked callback feature now.

File: lib/matplotlib/colorbar.py
Comment (QuLogic) [inline lib/matplotlib/colorbar.py:481]: What is this for?

File: lib/matplotlib/tests/test_colors.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_colors.py:1417]: Since this does nothing otherwise, `increment = unittest.mock.Mock(return_value=None)`

File: lib/matplotlib/tests/test_colors.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_colors.py:1423]:
```suggestion
    assert increment.call_count == 0
```
etc.

File: lib/matplotlib/colorbar.py
Comment (greglucas) [reply lib/matplotlib/colorbar.py:474]: When we go through the new update path we try to update the minorlocator, but it complains about no attribute exists, so we should just make it available in __init__ right away.

File: lib/matplotlib/tests/test_colors.py
Comment (greglucas) [reply lib/matplotlib/tests/test_colors.py]: Great suggestion, thanks!

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/tests/test_image.py
Comment (tacaswell) [reply lib/matplotlib/tests/test_image.py:1020]: This is fun second-order behavior change!

- In https://github.com/matplotlib/matplotlib/pull/6700 we stop passing vmin/vmax through `float` on init because it stripped precision of float128
- this broke yt that was expecting this to strip off units so we got https://github.com/matplotlib/matplotlib/pull/6700 to add `_sanitize_extrema` which _gently_ passes things through float
- this got updated in #12478 due to changes in numpy to deprecate np.asscalar 
- this PR now passes _all_ setting of vmin/vmax through `_sanitize_extrema`  (instead of previously just on init)
- `np.ma.masked.item() == 0` but `float(np.ma.masked) is np.nan`

Previously if you did 

```python
im = plt.imshow(np.full((10, 10), np.nan), vmin=np.ma.masked, vmax=np.ma.masked)
assert im.norm.vmin == 0
assert im.norm.vmax == 0
```

but if you did 

```python
im = plt.imshow(np.full((10, 10), np.nan)
assert im.norm.vmin is np.ma.masked
assert im.norm.vmax is np.ma.masked
```

(`==` is nan-like with `np.ma.masked`).

I would argue that this is a bug, in the sense that explicitly passing the `np.max(data)` and letting us auto-computer the max gave different results.  

We also say that if vmin == vmax the result of `__call__` will be 0.

This will lead to a subtle behavior change in the case where:
 - the user passed us all nan
 - let us infer the vmin/vmax
 - later updates the array with non-nan values
  
Previously the user would get all "bad" color, now they will get the "0" color.  Given the other difference in behavior of explicitly passing vmin/vmax, the agreement with "if vmin/vmax match -> go to 0", this feeling like a very cornery corner case, and the results already being useless from a communications point of view, I think we can leave this as-is and do not need a further API change note.

**Review by tacaswell — APPROVED** [review]

**Pull request merged**
